MGDSTRM-72: first pass at kafka SLI metrics patterns

jeremyary opened the pull request:
For @ppatierno @pb82 review/place for conversation on metrics filters

jeremyary commented:
To copy over some thoughts from the Jira issue:
We haven't named any SLI's containing zookeeper metrics, so nothing for that section yet. Some others metrics in spreadsheet are calculations rather than just metrics, so I've exposed the metrics necessary for those calcs. I'm not sure I can do sum/percentile/etc in patterns, so looking for that. Other SLI metrics come from k8s & strimzi operator, so no filtering/pattern work to be done for those, and a few are still WIP upstream, so will require a circle back to complete.

jeremyary commented:
Zookeeper patterns updated to match provided SLI's.

jeremyary commented:
IMO it would be worthwhile for someone to look over the output from a pod from each type (or try them yourself) and see what you think regarding output inclusion/format:

https://gist.github.com/jeremyary/e59849e1664e8d86fd7e1aa800bc70f1

david-martin commented:
@jeremyary #9 looks like it will land soon.
Some dashboards are created by 1 of the targets added in that PR https://github.com/RHCloudServices/kafka-monitoring-stuff/pull/9/files#diff-f770bff7af7663b520ef320a9503b675R1

Before landing these changes to the Kafka CR, let's check if the queries used by those dashboards reference the correct metric patterns (and update either the pattern or the dashboard)

ppatierno commented on install/resources/cluster/kafka.yaml:
```diff
@@ -23,147 +23,79 @@ spec:
-        labels:
-          quantile: "0.$4"
+        # WORKING, results are per pod, all show == 0
+        - pattern: kafka.controller<type=KafkaController, name=OfflinePartitionsCount>
```
I think that the domani (i.e. `kafka.controller` in this case) should be always part of the metrics name in order to understand immediately from where the metric is coming.
So I would suggest that in this specific case the renaming should be like `kafka_controller_offline_partitions`.
We should do the same for all the others as well.

ppatierno replied:
I would also suggest to have the attribute name to be specified so in this case it should be:

`kafka.controller<type=KafkaController, name=OfflinePartitionsCount><>Value`

it's useful in case an mbean has more than one attribute.

ppatierno commented on install/resources/cluster/kafka.yaml:
```diff
@@ -23,147 +23,79 @@ spec:
+          type: GAUGE
+        # WORKING, but topic list may grow quite large since it's per partition
+        - pattern: kafka.cluster<type=Partition, name=AtMinIsr, topic=(\w+), partition=(.*)><>
+          name: kafka_at_min_isr_partitions_by_topic
```
do we really need the kind of grouping "by XXX" in the metric name? I think that grouping cames later when you query a metric and you could even group by partition (even if it could not making any sense).
I would leave the grouping out of names.

ppatierno commented on install/resources/cluster/kafka.yaml:
```diff
@@ -23,147 +23,79 @@ spec:
+          type: GAUGE
+        # WORKING, but topic list may grow quite large since it's per partition
+        - pattern: kafka.cluster<type=Partition, name=UnderMinIsr, topic=(\w+), partition=(.*)><>
+          name: kafka_under_min_isr_partitions_by_topic
```
ditto as above about grouping out of the name.

ppatierno commented on install/resources/cluster/kafka.yaml:
```diff
@@ -23,147 +23,79 @@ spec:
+        # WORKING, results are per pod
+        - pattern: kafka.server<type=BrokerTopicMetrics, name=BytesOutPerSec>
+          name: kafka_cluster_throughput_out
+          type: GAUGE
```
for bytes in and out I would use a metric name closer to the "name" itself. I mean using "bytes" instead of "throughput". It's about SLI & SLO based on this metric to get that the metric is useful for throughput but I would be more generic in the metric name itself.

ppatierno commented on install/resources/cluster/kafka.yaml:
```diff
@@ -23,147 +23,79 @@ spec:
+        - pattern: kafka.server<type=BrokerTopicMetrics, name=BytesOutPerSec>
+          name: kafka_cluster_throughput_out
+          type: GAUGE
+        # WORKING BUT... results are per-request, need to figure out percentile or threshold
```
We can do in the following way, getting the attribute in the metrics name and it will be main, avg, 50thPercentile, things like that:

```yaml
- pattern: kafka.network<type=RequestMetrics, name=TotalTimeMs, request=(\w+)><>(\w+)
          name: kafka_network_request_latency_$2
          type: GAUGE
          labels:
            request: "$1"
```

for example

```
kafka_network_request_latency_50thpercentile{request="DeleteGroups",} 0.0
kafka_network_request_latency_max{request="CreatePartitions",} 0.0
```

ppatierno commented on install/resources/cluster/kafka.yaml:
```diff
@@ -23,147 +23,79 @@ spec:
+          type: GAUGE
+          labels:
+            request: "$1"
+        # WORKING BUT... getting non-zero for a single pod, zero on others, doesn't seem right? only come from active controller maybe?
```
yes exactly, this metric comes from the mbean related to the active controller which is always just one.

ppatierno commented:
@david-martin @jeremyary I had a quick first pass to the dashboards we have in #9 and because they are coming from upstream, the patterns provided by this PR are going to break the dashboards. Let me summarize how upstream we defined the patterns compared to the Kafka JMX mbean.

So, an mbean has the following format:

```
domain<beanpropertyName1=beanPropertyValue1, beanpropertyName2=beanPropertyValue2, ...><key1, key2, ...>attrName: value
```

Where in our specific case, Kafka mbeans have a "key" part empty, so it's just <> and the main propertiesies are "name" and "type" plus others (i.e. "topic", "partition" and so on depending on the metric) ... so something like that:

```
domain<type=value, name=value, .....><>attrName
```

Upstream patterns are defined in order to have always this format for the metric name:
```
<domain>_<type>_<name>
```

Then other properties are sometimes in the name, other times in the labels. It's different case by case.
For example, an mbean like:

```
kafka.controller<type=KafkaController, name=ActiveControllerCount><>Value
```

becomes:

```
kafka_controller_kafkacontroller_activecontrollercount{ ... }
```

Leaving this kind of format would help to reuse the dashboards much more and I think it's a good format. 
Wdyt?

jeremyary commented:
> Leaving this kind of format would help to reuse the dashboards much more and I think it's a good format.
> Wdyt?

I started off just using the names that we gave the SLI's fully expecting it would be altered. IMO, the upstream naming pattern is a bit abstract and hard to relate to the value of the metric sometimes, but absolutely easier to trace back to the source & understand exactly what's producing it, so I see pros/cons of both. I'm ok with implementing the upstream format if we'd prefer avoiding the rework to dashboards among other benefits, I just assumed that we'd be doing that anyway and building our own while the upstream dashboards were most 'getting started' points.

jeremyary commented:
I've updated PR to include the suggested changes from @ppatierno as well as the new SLI's added around perf metrics. I've also updated this gist to reflect the new filtered kafka/zk metrics from each pod type:

https://gist.github.com/jeremyary/e59849e1664e8d86fd7e1aa800bc70f1

I have not made any changes regarding Paolo's last discussion about naming pattern upstream vs. current yet, but I did want to go ahead and push up the new SLI's. Once we formalize that decision, I'll update.

jeremyary commented:
Also, just to jot down somewhere with more permanence, what I meant by metrics that don't require patterns are those on our SLI list that are coming from alternate pod sources (strimzi operator, kubernetes)  outside of the `kafka:` and `zookeeper:` CR config... I believe those to include:

- container_network_receive_bytes_total
- container_network_transmit_bytes_total
- container_memory_usage_bytes
- container_cpu_user_seconds_total
- container_start_time_seconds{container="strimzi-cluster-operator"}
- kubelet_volume_stats_available_bytes
- kubelet_volume_stats_capacity_bytes
- kube_persistentvolumeclaim_status_condition
- kube_persistentvolumeclaim_status_phase
- kube_pod_status_ready
- kube_pod_status_phase
- strimzi_reconciliations_successful_total{kind="Kafka",}
- strimzi_resource_state{kind="Kafka", name="my-cluster", resource-namespace="my-kafka-namespace"}

ppatierno commented:
> the upstream naming pattern is a bit abstract and hard to relate to the value of the metric sometimes

what do you mean by that? I see that you have changed the PR providing the entire <domain> part in the metric name but not adding <type> ... about <name> part it's almost the same between current PR and upstream.
i.e.

`kafka_cluster_at_min_isr_partitions`
vs
`kafka_cluster_partition_atminisr`

or

`kafka_server_under_min_isr_partitions`
vs
`kafka_server_replicamanager_underminisrpartitioncount`

I was suggesting using the pattern with `<domain>_<type>_<name>` because at the end SRE and engineers have to correlate the metric with the corresponding mbean in Kafka to understand what JMX metric was resolved. It could be done checking the pattern used in the JMX exporter configuration but would be much easier if reflecting all the information (domain, type, and name of the mbean) just in the metric name.
Just my 2 cents.

david-martin commented:
Summarising my comments from offline.

`under_min_isr_partitions` is more readable than `underminisrpartitioncount` and is inline with other metrics sources we use (e.g. kube-state & kubelet)

Having the bean name in the metric seems fine.
An SRE should *not* need to know the bean name or have to lookup things in code.
An SRE should *not* need to engage with an AMQ Streams engineer/expert to get a basic understanding of what each metric is for. However, if/when they do, it's trivial to determine how the metric maps to the underlying code in a bean.
An SRE *should* have a reference doc for what each metric means

david-martin commented:
on hold pending downstream discussion

jeremyary commented:
- applied the naming pattern @david-martin requested
- some patterns have been modified and others added (and noted as such) to support the upstream dashboards currently in use

@ppatierno @pb82 I've looked at the cluster dashboards with these patterns applied and resolved any issues I could identify, but would be helpful if you guys could apply and take a look since you have domain knowledge and/or know what's reasonably normal to see. Thanks!

ppatierno reviewed (approved):
LGTM. I just fixed the ZooKeeper alive connections in the corresponding dashboard.
